File: backend/app/schemas/scorecard.py
from __future__ import annotations
from typing import Any, Optional
from datetime import date, datetime
from pydantic import BaseModel


class ScorecardKPIValue(BaseModel):
    value: Optional[float]
    contributing_ids: list[str]


class ScorecardDimension(BaseModel):
    score: Optional[float]
    trend: Optional[str]
    kpis: dict[str, ScorecardKPIValue]


class ScorecardTopMove(BaseModel):
    assessment_id: str
    signal_id: str
    title: str
    movement_score: int
    signal_class: str


class ScorecardRiskFlag(BaseModel):
    assessment_id: str
    capability_key: str
    movement_strength: str
    title: str


class ScorecardBenchmarkPosition(BaseModel):
    rank: int
    percentile: float
    total_competitors: int


class ScorecardRead(BaseModel):
    id: str
    company_id: str
    period_type: str
    period_start: date
    period_end: date
    generated_at: datetime
    overall_score: Optional[float]
    overall_trend: Optional[str]
    dimension_scores: dict[str, ScorecardDimension]
    top_capabilities: list[dict[str, Any]]
    top_moves: list[ScorecardTopMove]
    risk_flags: list[ScorecardRiskFlag]
    watchpoints: list[str]
    benchmark_position: Optional[ScorecardBenchmarkPosition]
    contributing_assessment_ids: list[str]
    is_current: bool
    scorecard_version: Optional[str]
    routing_version: Optional[str]

    model_config = {"from_attributes": True}


class ScorecardHistoryItem(BaseModel):
    id: str
    overall_score: Optional[float]
    overall_trend: Optional[str]
    generated_at: datetime
    scorecard_version: Optional[str]

    model_config = {"from_attributes": True}


class ScorecardExplainAssessment(BaseModel):
    assessment_id: str
    signal_id: str
    title: str
    movement_score: int
    signal_class: str


class ScorecardExplainDimension(BaseModel):
    dimension: str
    score: Optional[float]
    dimension_weight: float
    effective_weight: float
    weighted_contribution: Optional[float]
    assessment_count: int
    top_contributing_assessments: list[ScorecardExplainAssessment]
    kpi_detail: dict[str, ScorecardKPIValue]


class ScorecardExplain(BaseModel):
    overall_score: Optional[float]
    dimension_breakdown: list[ScorecardExplainDimension]
    null_dimensions: list[str]
    score_formula: str
    routing_version: Optional[str]
    scorecard_version: Optional[str]


class ScorecardRecomputeAck(BaseModel):
    status: str
    company_slug: str
    recomputed_periods: list[str]
    scorecard_ids: dict[str, str]
    generated_at: datetime


class BenchmarkScorecardItem(BaseModel):
    company_id: str
    slug: str
    name: str
    overall_score: Optional[float]
    rank: int
    percentile: float
    dimension_scores: dict[str, ScorecardDimension]
    overall_trend: Optional[str]
    scorecard_version: Optional[str]


class BenchmarkScorecardView(BaseModel):
    items: list[BenchmarkScorecardItem]
    total: int
    page: int
    page_size: int
    pages: int
    period_type: str
    capability_leaders: dict[str, dict[str, Any]]
    highest_momentum: Optional[dict[str, Any]]
    threat_flags: list[dict[str, Any]]

File: frontend/src/types/scorecard.ts
export type ScorecardPeriodType = '30d' | '90d' | '180d';
export const SCORECARD_PERIOD_TYPES: ScorecardPeriodType[] = ['30d', '90d', '180d'];

export interface ScorecardKPIValue {
  value: number | null;
  contributing_ids: string[];
}

export interface ScorecardDimension {
  score: number | null;
  trend: 'rising' | 'stable' | 'declining' | null;
  kpis: Record<string, ScorecardKPIValue>;
}

export interface ScorecardTopMove {
  assessment_id: string;
  signal_id: string;
  title: string;
  movement_score: number;
  signal_class: string;
  published_at: string | null;
}

export interface ScorecardRiskFlag {
  assessment_id: string;
  signal_id: string;
  capability_key: string;
  movement_strength: string;
  title: string;
}

export interface ScorecardBenchmarkPosition {
  rank: number;
  percentile: number;
  total_competitors: number;
}

export interface CompetitorScorecard {
  id: string;
  company_id: string;
  period_type: ScorecardPeriodType;
  period_start: string;
  period_end: string;
  generated_at: string;
  overall_score: number | null;
  overall_trend: 'rising' | 'stable' | 'declining' | null;
  dimension_scores: Record<string, ScorecardDimension>;
  top_capabilities: Array<{ capability_key: string; score: number | null }>;
  top_moves: ScorecardTopMove[];
  risk_flags: ScorecardRiskFlag[];
  watchpoints: string[];
  benchmark_position: ScorecardBenchmarkPosition | null;
  contributing_assessment_ids: string[];
  is_current: boolean;
  scorecard_version: string | null;
  routing_version: string | null;
}

export interface ScorecardHistoryItem {
  id: string;
  overall_score: number | null;
  overall_trend: 'rising' | 'stable' | 'declining' | null;
  generated_at: string;
  scorecard_version: string | null;
}

export interface ScorecardExplainAssessment {
  assessment_id: string;
  signal_id: string;
  title: string;
  movement_score: number;
  signal_class: string;
}

export interface ScorecardExplainDimension {
  dimension: string;
  score: number | null;
  dimension_weight: number;
  effective_weight: number;
  weighted_contribution: number | null;
  assessment_count: number;
  top_contributing_assessments: ScorecardExplainAssessment[];
  kpi_detail: Record<string, ScorecardKPIValue>;
}

export interface ScorecardExplain {
  overall_score: number | null;
  dimension_breakdown: ScorecardExplainDimension[];
  null_dimensions: string[];
  score_formula: string;
  routing_version: string | null;
  scorecard_version: string | null;
}

export interface BenchmarkScorecardItem {
  company_id: string;
  slug: string;
  name: string;
  overall_score: number | null;
  rank: number;
  percentile: number;
  dimension_scores: Record<string, ScorecardDimension>;
  overall_trend: string | null;
  scorecard_version: string | null;
}

export interface BenchmarkScorecardView {
  items: BenchmarkScorecardItem[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
  period_type: ScorecardPeriodType;
  capability_leaders: Record<string, { company_slug: string; score: number }>;
  highest_momentum: { company_slug: string; mom_period_delta: number } | null;
  threat_flags: Array<{ company_slug: string; capability: string; movement_strength: string }>;
}

export interface ScorecardRecomputeAck {
  status: string;
  company_slug: string;
  recomputed_periods: string[];
  scorecard_ids: Record<string, string>;
  generated_at: string;
}
